src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -0,0 +1,328 @@
+                "node, but it is on the right-hand-side.")
+
+        # Contains a range node
+        if node.walk(Range):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node should not be or contain a Range node as it "
+                "should be single valued.")
+
+        # Capture loop iterator symbols in order
+        iterator_symbols = []
+        location = node.parent.parent
+        while (isinstance(location.parent, Schedule) and
+               isinstance(location.parent.parent, (Loop, NemoKern))):
+            location = location.parent.parent
+            if isinstance(location, Loop):
+                iterator_symbols.append(location.variable)
+
+        # Index contains a loop iterator
+        for reference in node.walk(Reference):
+            if reference.symbol in iterator_symbols:
+                raise TransformationError(
+                    "Error in NemoArrayAccess2LoopTrans transformation. The "
+                    "supplied node should not be or contain a loop iterator, "
+                    "it should be single valued.")
+
+        # An index before this one has more than one loop iterator
+        array_access_iterators = []
+        for index in node.parent.children[:node.position]:
+            # Find any iterators in this index
+            index_iterators = [
+                my_index.symbol for my_index in index.walk(Reference)
+                if my_index.symbol in iterator_symbols]
+            iterator_names = [my_index.name for my_index in index_iterators]
+            if len(index_iterators) > 1:
+                raise TransformationError(
+                    "Only a single iterator per dimension is supported by "
+                    "this transformation, but found '{0}'."
+                    "".format(iterator_names))
+            if index_iterators:
+                array_access_iterators.append(index_iterators[0])
+
+        # Same iterator used in more than one loop index
+        if len(iterator_symbols) < len(array_access_iterators):
+            raise TransformationError(
+                "The same iterator is used in more than one loop index "
+                "'{0}' which is not supported by this transformation.".format(
+                    [iterator.name for iterator in array_access_iterators]))
+
+        # Iterators before this index are not in loop index order
+        for idx, array_access_iterator in enumerate(array_access_iterators):
+            if array_access_iterator != iterator_symbols[idx]:
+                raise TransformationError(
+                    "In array '{0}' expected iterator '{1}' at index '{2}' "
+                    "but found '{3}'.".format(
+                        node.parent.name, iterator_symbols[idx].name,
+                        idx, array_access_iterator.name))
+
+        # Indices on lhs and rhs array accesses are not the same
```

Comment:
@hiker is introducing/tidying functionality for precisely this kind of check. That should be used here.

Reply:
Am I not using this functionality with the `math_equal()` method? @hiker, could you please advise.

Reply:
I am not sure what is being suggested here given our recent discussion about having a `tree_compare()` method and limiting the scope of use for `math_equal()`. If I created an issue about `tree_compare()` and reference it here and kept `math_equal()` for the time being would that be acceptable?

Reply:
Happy for @hiker to advise but adding new functionality to `array_accesses_consistent` sounds like a different issue. I can use this function instead and then add an issue if that is what you both think is best. However, I thought we were saying that it should be valid for `math_equal` to be used to compare expressions, so I am confused as to which one should use and when.

Reply:
My existing code does not really apply here, it only checks that an index variable is always used at the same dimension (E.g. a(i) and a(i+1) would be consistent), while here you also need the check the value. I will be doing something similar in the future, but at this stage nothing is ready. It might be worth considering if this functionality could be useful elsewhere, then it should go into the dependency tools.
Otherwise using `math_equal` here seems the right thing to do, it would accept `i+j` and `j+i`, and when we support symbolic maths would be replaced to be more general.
If you want me to have a closer look at the code, and check if there is anything in dependency tools (or so) that could be useful, let me know - and I need a bit more time than I have right now to understand what is happening here.

Reply:
Thanks @hiker, I'm happy to continue to use `math_equal` and agree that it feels like the right thing to use here.